ls1intum/Artemis
# `Programming exercises`: Fix an issue in the score calculation in case of many SCA issues

### Checklist
#### General
#### Server

### Motivation and Context
On production, we encountered a student who got a too high point deduction for SCA issues (above the limit specified in the maximum SCA deduction setting)

### Description
On many places in Artemis, it is assumed that you simply can sum up the credits of all feedback elements in order to get the score. To support this, Artemis already adjusts the credits of SCA feedback items in case the maximum value for the category got exceeded.

However, if the sum of all categories exceeded the limit specified by the exercise, the penalty was locally reduced but the feedback items were not changed accordingly.


Now, when a category would exceed the maximum penalty (if overallPenaltyPoints + categoryPenaltyPoints > maxExercisePenaltyPoints), the points for this category get reduced. This change is also reflected in all the related feedback items:

```
┌───────────────────────────────────────────────────────────────┐
│ maxExercisePenaltyPoints                                      │
└───────────────────────────────────────────────────────────────┘
┌──────────────────────────────────────┬────────────────────────────────────────────┐
│ overallPenaltyPoints                 │ categoryPenaltyPoints                      │
└──────────────────────────────────────┴────────────────────────────────────────────┘
                                       ┌────────────────────────┐
 ==>                                   │ categoryPenaltyPoints  │
                                       └────────────────────────┘
```

### Steps for Testing
Prerequisites:
- 1 Tutor
- 1 Student
- 1 Programming Exercise with SCA enabled and a maximum penalty set, also manual correction active

1. Student: Submit a solution that generates many SCA penalties (e.g. a lot of unused imports, unused attributes and methods)
2. Student: View your result and check that the limit got applied.
3. Tutor: After the due date, submit a manual correction.
4. Student: Verify that the manual correction got a correctly calculated score.

### Review Progress

#### Code Review
#### Manual Tests

### Test Coverage
changes covered
